### PR TITLE
Storages: simplify read path

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #6233

Problem Summary:

### What is changed and how it works?

Simplify read path

Before:

image

image

After:

image

image

Those changes bring two advantages:

1. It is more convenient for us to add functions like `loadVectorIndex` for both delta and stable.
2. It is more convenient for us to calculate user read bytes.

This PR will pick to cse.

```commit-message
Storages: simplify read path
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/SkippableBlockInputStream.cpp [hunk 1]
```diff
@@ -45,6 +45,14 @@ ConcatSkippableBlockInputStream<need_row_id>::ConcatSkippableBlockInputStream(
     current_stream = children.begin();
 }
 
+template <bool need_row_id>
+void ConcatSkippableBlockInputStream<need_row_id>::appendChild(SkippableBlockInputStreamPtr input, size_t rows)
+{
+    children.push_back(input);
+    this->rows.push_back(rows);
```
Comment (CalvinNeo) [inline dbms/src/Storages/DeltaMerge/SkippableBlockInputStream.cpp:52]: Why not just change the name of the parameter?

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/SkippableBlockInputStream.cpp]: Fixed

**Review by CalvinNeo — APPROVED** [review]
lgtm

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 2]
```diff
@@ -3243,32 +3245,38 @@ BlockInputStreamPtr Segment::getBitmapFilterInputStream(
...
-    return std::make_shared<BitmapFilterBlockInputStream>(
-        columns_to_read,
-        stable_stream,
-        delta_stream,
-        segment_snap->stable->getDMFilesRows(),
-        bitmap_filter,
-        dm_context.tracing_id);
+    auto stream = std::dynamic_pointer_cast<ConcatSkippableBlockInputStream<NeedRowID>>(stable_stream);
+    stream->appendChild(persisted_files_stream, persisted_files->getRows());
+    stream->appendChild(mem_table_stream, memtable->getRows());
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Segment.cpp:3274]: `persisted_files->getRows()` and `memtable->getRows()` may not be the actual rows when there are ColumnFileBig but only a part of the key-range of the DMFile is covered for this segment.
Does it affect the correctness?

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:3275]: No, when building the bitmap, all physical rows will be included.

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 3]
```diff
@@ -3308,32 +3310,39 @@ BlockInputStreamPtr Segment::getBitmapFilterInputStream(
...
+        read_tag);
 
-    return std::make_shared<BitmapFilterBlockInputStream>(
-        columns_to_read,
-        stable_stream,
-        delta_stream,
-        segment_snap->stable->getDMFilesRows(),
-        bitmap_filter,
-        dm_context.tracing_id);
+    auto stream = std::dynamic_pointer_cast<ConcatSkippableBlockInputStream<NeedRowID>>(stable_stream);
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:3337]: `ConcatSkippableBlockInputStream` accepts a template argument `need_row_id`

File: dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.cpp [hunk 4]
```diff
@@ -207,14 +194,6 @@ size_t ColumnFileSetReader::readRows(
         }
     }
 
-    if (auto delta_bytes = columnsSize(output_columns) - bytes_before_read; delta_bytes > 0)
-    {
-        if (row_ids == nullptr)
-            lac_bytes_collector.collect(delta_bytes);
-        if (likely(context.scan_context))
-            context.scan_context->user_read_bytes += delta_bytes;
-    }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.cpp:216]: Where do we add the bytes read of delta layer to the `user_read_bytes` after these code changes?

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileSetReader.cpp:216]: ConcatSkippableBlockInputStream

**Pull request merged**